ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx

```diff
@@ -0,0 +1,197 @@
+  const rootItem = itemsByPath?.[rootPath];
+  const [data, setData] = useState(null);
+  const [itemMenu, setItemMenu] = useSpreadState<Menu>({
+    path: null,
+    anchorEl: null,
+    loaderItems: null
+  });
+  const nodesByPathRef = useRef({});
+  const fetchingPathsRef = useRef([]);
+
+  const dispatch = useDispatch();
+  useMount(() => {
+    dispatch(
+      pathNavigatorTreeInit({
+        id,
+        path: rootPath,
+        excludes,
+        limit
+      })
+    );
+  });
+
+  useEffect(() => {
+    if (rootItem) {
+      const rootNode = {
+        id: rootItem.path,
+        name: rootItem.label,
+        children: [{ id: 'loading' }]
+      };
+      nodesByPathRef.current[rootItem.path] = rootNode;
+      setData(rootNode);
+    }
+  }, [rootItem]);
+
+  useEffect(() => {
+    const nextFetching = [];
+    fetchingPathsRef.current.forEach((path) => {
+      if (childrenByParentPath[path]) {
+        nodesByPathRef.current[path].children = [];
+        childrenByParentPath[path].forEach((childPath) => {
+          const node = {
+            id: childPath,
+            name: itemsByPath[childPath].label,
+            children: [{ id: 'loading' }]
+          };
+          nodesByPathRef.current[path].children.push(node);
+          nodesByPathRef.current[childPath] = node;
+        });
+
+        setData({ ...nodesByPathRef.current[rootPath] });
+      } else {
+        nextFetching.push(path);
+      }
+    });
+    fetchingPathsRef.current = nextFetching;
+  }, [data, rootPath, childrenByParentPath]);
+
+  const onChangeCollapsed = (collapsed) => {};
+
+  const onLabelClick = (path: string) => {
+    if (isNavigable(itemsByPath[path])) {
+      console.log('preview: ', path);
+    } else {
+      onIconClick(path);
+    }
+  };
+
+  const onIconClick = (path: string) => {
+    // If the path is already expanded, pathNavigatorTreeUpdate should collapsed it
+    if (state.expanded.includes(path)) {
+      dispatch(
+        pathNavigatorTreeUpdate({
+          id,
+          expanded: state.expanded.filter((expanded) => expanded !== path)
+        })
+      );
+    } else {
+      // if: the path does not exist on childrenByParentPath, the children was not fetched
+      // else: the tree should expand the selected node
+      if (!childrenByParentPath[path]) {
+        fetchingPathsRef.current.push(path);
+        dispatch(
+          pathNavigatorTreeFetchPathChildren({
+            id,
+            path
+          })
+        );
+      } else {
+        dispatch(
+          pathNavigatorTreeUpdate({
+            id,
+            expanded: [...state.expanded, path]
+          })
+        );
+      }
+    }
+  };
+
+  const onOpenItemMenu = (element: Element, path: string) => {
+    setItemMenu({
+      path,
+      anchorEl: element,
+      loaderItems: getNumOfMenuOptionsForItem(itemsByPath[path])
+    });
+  };
+
+  const onCloseItemMenu = () => setItemMenu({ ...itemMenu, path: null, anchorEl: null });
```

Comment:
no need to spread; you only need to set anchorEl to null?

Reply:
it is to keep the number of loaderItems when the menu is being hidden